### ISSUE TITLE
Provide instructions for building/developing LSP4J itself

### ISSUE THREAD

**Issue by PerMildner** [eclipse/lsp4j#344]
It is not clear how to hack LSP4J itself (using Eclipse).

I want to implement and contribute a fix for a LSP4J bug but I can not find any information about how to develop/build LSP4J itself using Eclipse.

(I tried Import "Projects from Git (with smart import)" from https://github.com/eclipse/lsp4j.git but that produced several broken projects.)

**Comment by spoenemann** [comment]
1. Install [Xtend](https://www.eclipse.org/xtend/download.html).
2. Import the projects with Buildship: _Import_ &rarr; _Gradle_ &rarr; _Existing Gradle Project_.

**Comment by PerMildner** [comment]
Thanks. I need a little more details since I know nothing about Gradle and Buildship (I need "for Dummies"-level instructions...).

The Import->"Existing Gradle Project" wizard wants me to "Specify the root directory of the Gradle project to import.", how do I get that from https://github.com/eclipse/lsp4j.git, preferably from within Eclipse?

**Comment by spoenemann** [comment]
Clone the repository to a local directory (e.g. with [EGit](https://www.eclipse.org/egit/)) and then use that as root directory in the import wizard.

**Comment by PerMildner** [comment]
What would be a suitable way to "Clone the repository to a local directory" with EGit?
When I try Import->"Projects from Git" it insists on creating a Project and requires that I know what kind of project to create but I do not know that.

(Sorry for being so obnoxious but I really do not know the answers to the above questions and this is why it would be good with some "How to contribute" information on the LSP4J site.)

**Comment by jonahgraham** [comment]
I think what @spoenemann meant was you could run `git clone https://github.com/eclipse/lsp4j.git` in a terminal. You may be able to use egit too, but the flow may be awkward.

I have written the detailed instructions in https://github.com/eclipse/lsp4j/blob/master/Contributing.md - but it perhaps assumes more knowledge than it should so I can revisit it to make step 2 clearer.

**Comment by jonahgraham** [comment]
PS have a look at issue #199 about unexpected changes to some of the generated files you might encounter.

**Comment by PerMildner** [comment]
Now I feel like an idiot:-)

I read the README.md and https://projects.eclipse.org/projects/technology.lsp4j/developer but found nothing about building. I never looked further.

The Contributing.md  seems to cover everything I need (except, perhaps, that the first few pages are boring legalese so I may have never gotten to the how-to-build part).

EGit indeed seems useless here, a plain git clone from the command line and the instructions in Contributing.md  seems to cover everything I need.

Thanks for all the help.

**Comment by PerMildner** [comment]
Except for perhaps making Contributing.md more visible (e.g. mention it in README.md and https://projects.eclipse.org/projects/technology.lsp4j/developer) it seems my issue is more or less a duplicate of the already fixed #196.

**Comment by jonahgraham** [comment]
> Except for perhaps making Contributing.md more visible

Done See #345

**Comment by jonahgraham** [comment]
> EGit indeed seems useless here, a plain git clone from the command line and the instructions in Contributing.md seems to cover everything I need.

While I use egit extensively, for whatever reason I almost always clone from the command line. You can indeed use egit to clone the repo just fine, but gradle/buildship is not integrated fully so you can't import the projects this way. Make sure you leave the import projects unchecked:

![image](https://user-images.githubusercontent.com/679236/62404050-83d7f680-b55f-11e9-85b8-73de664e2c3e.png)

I have filed [Bug 549760](https://bugs.eclipse.org/bugs/show_bug.cgi?id=549760) requesting Buildship to improve the flow so that LSP4J can be more easily imported using egit.

Thank you for persisting and I and the rest of the community are looking forward to your fix. Please let us know if there is anything else we can do to unblock you.

**Comment by PerMildner** [comment]
The bug I wanted to report and fix is #346.
